### DISCUSSION TITLE
how to handle t.Optional[SomeClass] for dataclasses

### DISCUSSION THREAD

**Discussion by jjmachan** [microsoft/pyright#7098] [Q&A]
I'm trying to create a class with dataclass like this 

```py
@dataclass
class A:
	attr: Optional[str] = None

	def __post_init__(self):
		if self.attr is None:
			# fix it in post init so the attr will never be none later down the code
	def use_attr(self):
		self.attr.format() # error because type checks assume it can be none and str and report that None has no function format()
```

right now I'm using an assert statement to force pyright to detect this but how can I improve this. I dont' want to have to use assert/check_for_none every time I want to use attr

any help highly appreciated

**Comment by erictraut** [comment]
This approach won't work because you're declaring that `attr` accepts both `str` and `None`. The fact that you've assigned a `str` value in `__post_init__` doesn't prevent it from being reassigned to `None` in some other code path. If your intent is for `attr` to always be `str`, then it should be declared as such.

Rather than fix it in post-init, you could use a default value.

```python
@dataclass
class A:
    attr: str = ""

    def use_attr(self):
        self.attr.format()
```

**Reply by jjmachan** [reply]
thanks for replying Eric :)

I see that the solution would be to create default value of the same type but what if the type in question is a complicated class and so creating a default value might be complicated right

```py
# importing some class from a library
from langchain.llms import OpenAI, BaseLLM

@dataclass
class Evolution:
    llm: BaseLLM = # what can I do here?
```

I can't create an instance of OpenAI because it will do some other checks, so what is a good default? should I create something specifically to be used as default?

**Reply by erictraut** [reply]
I'm not sure I understand. You are able to initialize it in your `__post_init__` method, right? How would you do it there?

Another thought... Do you really need a dataclass here? Would a regular class make sense?

If you really want to use a dataclass, one potential solution is to use an `InitVar` and a `field` with `init` set to `False`.

```python
from dataclasses import InitVar, dataclass, field

@dataclass
class A:
    # This adds `_attr` to the `__init__` method but forwards its value to `__post_init__`
    _attr: InitVar[str | None] = None

    # This omits `attr` from the `__init__` method but allocates space for it in the dataclass
    attr: str = field(init=False)

    def __post_init__(self, _attr: str | None = None):
        self.attr = _attr or "Default"

    def use_attr(self):
        self.attr.format()

a = A()
print(a.attr)
```

**Reply by jjmachan** [reply]
let me see if I can explain it better. I understood that putting in a sensible default when initializing, would work perfectly but the issue is that for complex classes, it is hard to have sensible defaults. In a way, I use None as a default to work around the fact that there is no sensible default for the type I intend to use. default values for internal types like string are straightforward  but for more complex classes/types it's hard

Now I'm planning to do something like
```py
@dataclass
class B:
    attr: str = t.cast(str, None)
def __post_init__(self):
        if self.attr is None:
            # patch it with a new value
```
feel ugly but gets the job done. defeats the purpose of having types though sadly. but None is a much easier no-brainer default to have I feel.

what do you think?

**Reply by erictraut** [reply]
I'd avoid using `cast` because, as you said, it defeats the purpose of static type checking. It makes your code brittle and fragile.

Note that the dataclass `field` also accepts a `default_factory` parameter which accepts an arbitrary function. Perhaps this would address your issue?

```python
@dataclass
class B:
    @staticmethod
    def default_attr_factory() -> str:
        # This function can handle complex cases.
        return ""

    attr: str = field(default_factory=default_attr_factory)
```

**Reply by jjmachan** [reply]
I explore that too but might not work due to a different reason (and I think this could be design fault I have in my code).

So basically if the classes for my use case are initialised with default values (`None`s). Then at run time based on some other conditions, those values are replaced. This actually happens after which is why `default_factor` and `__post_init__` doesn't actually work.

so basically 
```py
@dataclass
class B:
    attr: str = t.cast(str, None)
def init_this(self, some_args):
        if self.attr is None:
            # patch it with a new value based or args

# init class with default
b = B()

# down the line we fill it and replace the default values
# now the type checker doesn't know I did all this
b.init_this(args)
```

this works because of some conditions in our codebase (setting precedence, have to wait for some conditions before init-ing some external library classes that are part of `B`) but this is what forced us to adopt the `init_this()` method.


I'm thinking of having a small runtime checker method in my class to check if the default values are set and raise an exception to work around the fact that the type checker might not pick up those bugs

**Reply by jjmachan** [reply]
btw thanks a lot for helping me brainstorm @erictraut ❤️

**Comment by Viicos** [comment]
I think this is a common problem with dataclasses with respect to typing support. The canonical example is:

```python
@dataclass
class MyClass:
    a: int
    b: int | None = None

    def __post_init__(self):
        if self.b is None:
            self.b = self.a + 1
```

And doesn't play well with type checkers, as `b` will always be `int`. In this case, `default_factory` can't be used because `b` depends on `a`.

**Reply by erictraut** [reply]
I don't think it's that common to have such dependencies between fields (such as `a` and `b` in your example). In rare cases where such dependencies exist, there are other approaches that avoid the problem and maintain good typing hygiene. For example, you could create an alternate constructor that accepts `None` for `b`.

```python
@dataclass
class MyClass:
    @staticmethod
    def create(a: int, b: int | None) -> "MyClass":
        return MyClass(a, b if b is not None else a + 1)

    a: int
    b: int
```

**Reply by Viicos** [reply]
Yes there are definitely workarounds to this (like the one you gave [here](https://github.com/microsoft/pyright/discussions/7098#discussioncomment-8217453)), but imo not really a big deal considering it's not really common.

This reminds me of [this interesting article](https://florimond.dev/en/posts/2018/10/reconciling-dataclasses-and-properties-in-python), where workarounds described to make properties work well with dataclasses.
